test/e2e/defaulting.go

```diff
@@ -0,0 +1,171 @@
+		}
+
+		if defaultKubeFedConfig == nil {
+			defaultKubeFedConfig = &v1beta1.KubeFedConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: qualifiedName.Name,
+				},
+				Spec: v1beta1.KubeFedConfigSpec{
+					// Set some values different from the defaults used. This
+					// ensures they are retained after defaulting is applied.
+					ControllerDuration: &v1beta1.DurationConfig{
+						AvailableDelay: &metav1.Duration{
+							Duration: defaults.DefaultClusterAvailableDelay + 11*time.Second,
+						},
+					},
+					FeatureGates: []v1beta1.FeatureGatesConfig{
+						{
+							Name:          string(features.PushReconciler),
+							Configuration: v1beta1.ConfigurationDisabled,
+						},
+					},
+					ClusterHealthCheck: &v1beta1.ClusterHealthCheckConfig{
+
+						Period: &metav1.Duration{
+							Duration: defaults.DefaultClusterHealthCheckPeriod + 11*time.Second,
+						},
+					},
+					Scope: apiextv1b1.ClusterScoped, // Required
+				},
+			}
+		}
+
+		// Ensure this gets set for every test even if the defaultKubeFedConfig
+		// is already initialized.
+		defaultKubeFedConfig.ObjectMeta.Namespace = qualifiedName.Namespace
+	})
+
+	It(fmt.Sprintf("%s is created by mutating admissing webhook and preserves values set by user", resourceName), func() {
+		By(fmt.Sprintf("Creating a %s defaulted by the mutating admission webhook", resourceName))
+		webhookDefaultedKubeFedConfig := defaultKubeFedConfig.DeepCopyObject().(*v1beta1.KubeFedConfig)
+		err := client.Create(context.TODO(), webhookDefaultedKubeFedConfig)
+		framework.ExpectNoError(err, fmt.Sprintf("Error creating default %q", *qualifiedName))
+
+		By(fmt.Sprintf("Creating a %s defaulted explicitly in this test", resourceName))
+		myDefaultedKubeFedConfig := defaultKubeFedConfig.DeepCopyObject().(*v1beta1.KubeFedConfig)
+		defaults.SetDefaultKubeFedConfig(myDefaultedKubeFedConfig)
+
+		By(fmt.Sprintf("Verifying the %s defaulted by the mutating admission webhook matches the one defaulted explicitly in this test", resourceName))
+		common.WaitForObject(tl, qualifiedName.Namespace, qualifiedName.Name, kubeFedConfigGetter, myDefaultedKubeFedConfig, framework.PollInterval, framework.TestContext.SingleCallTimeout, util.ObjectMetaAndSpecEquivalent)
+
+		if framework.TestContext.LimitedScope {
+			// Delete the KubeFedConfig we created since the kubefed system
+			// namespace will not be deleted when running namespace scoped.
+			framework.AddCleanupAction(func() {
+				fedConfig := webhookDefaultedKubeFedConfig
+				err := client.Delete(context.TODO(), fedConfig, fedConfig.Namespace, fedConfig.Name)
+				if err != nil && !apierrors.IsNotFound(err) {
+					f.Logger().Errorf("Error deleting %s %q: %v", resourceName, *qualifiedName, err)
+				}
+			})
+		}
+	})
+
+	// TODO(font): This test verifies the controller-manager does not fail
+	// early on an invalid KubeFedConfig that is created by the helm chart
+	// until https://github.com/kubernetes-sigs/kubefed/issues/983 is resolved.
```

Comment:
I'm not sure of this case.

Based on my understanding, there should no default settings in kubefed logic for `controller-manager` and there is no webhook for this case.  If `controller-manager` can be start (as this case expected), it means the system default (not setting by kubefed related logic but by API itself) can make `controller-manager` started up.

The `defaultKubeFedConfig` does not equal to helm installed `KubefedConfig`. And it works. It seems valid in starting a `controller-manager`. In the comments, it tells that this case can test invalid `KubeFedConfig` installed by helm. It seems not test invalid case. And invalid configuration should failed `controller-manager` start up. Any comments?

Reply:
I've updated the comment to hopefully make it more clear.